File: php/src/Generator.php
<?php

declare(strict_types=1);

namespace GeocitiesApp;

class Generator
{
    private static array $allPages = [
        ['key' => 'index',     'label' => '🏠 Home',      'file' => 'index.html'],
        ['key' => 'about',     'label' => '👤 About Me',   'file' => 'about.html'],
        ['key' => 'gallery',   'label' => '🖼️ Gallery',    'file' => 'gallery.html'],
        ['key' => 'guestbook', 'label' => '📖 Guestbook',  'file' => 'guestbook.html'],
        ['key' => 'links',     'label' => '🔗 Cool Links', 'file' => 'links.html'],
    ];

    private static string $templateDir = __DIR__ . '/../templates';

    public static function generate(string $outputDir, array $answers): array
    {
        $theme = Themes::get($answers['theme']);
        $pages = $answers['pages'] ?? [];

        @mkdir($outputDir . '/css', 0755, true);
        @mkdir($outputDir . '/js',  0755, true);

        $baseVars = [
            'SITE_NAME'       => $answers['site_name'],
            'AUTHOR_NAME'     => $answers['author_name'],
            'THEME_NAME'      => $theme['name'],
            'BG'              => $theme['bg'],
            'BG_PATTERN'      => $theme['bgPattern'],
            'BG_SIZE'         => $theme['bgSize'],
            'TEXT_COLOR'      => $theme['textColor'],
            'HEADING_COLOR'   => $theme['headingColor'],
            'LINK_COLOR'      => $theme['linkColor'],
            'LINK_HOVER'      => $theme['linkHover'],
            'ACCENT_COLOR'    => $theme['accentColor'],
            'BORDER_LIGHT'    => $theme['borderLight'],
            'BORDER_DARK'     => $theme['borderDark'],
            'PANEL_BG'        => $theme['panelBg'],
            'PANEL_BORDER'    => $theme['panelBorder'],
            'COUNTER_BG'      => $theme['counterBg'],
            'COUNTER_TEXT'    => $theme['counterText'],
            'TABLE_BORDER'    => $theme['tableBorder'],
            'TABLE_HEADER_BG' => $theme['tableHeaderBg'],
            'GLOW_COLOR'      => $theme['glowColor'],
            'NAV_ACTIVE_BG'   => $theme['navActiveBg'],
            'NAV_ACTIVE_TEXT' => $theme['navActiveText'],
            'CURSOR_EFFECT'   => $answers['cursor_effect'] ?? 'none',
            'FALLING_EFFECT'  => $answers['falling_effect'] ?? 'none',
            'WELCOME_ALERT'   => ($answers['welcome_alert'] ?? false) ? 'true' : 'false',
            'PLAY_MUSIC'      => ($answers['play_music'] ?? false) ? 'true' : 'false',
            'FAKE_HIGH_COUNT' => ($answers['fake_high_count'] ?? false) ? 'true' : 'false',
            'YEAR'            => (string) date('Y'),
        ];

        $css = file_get_contents(self::$templateDir . '/css/style.css');
        file_put_contents($outputDir . '/css/style.css', self::interpolate($css, $baseVars));

        $js = file_get_contents(self::$templateDir . '/js/main.js');
        file_put_contents($outputDir . '/js/main.js', self::interpolate($js, $baseVars));

        $pagesToGenerate = [['key' => 'index', 'file' => 'index.html']];
        foreach ($pages as $p) {
            $pagesToGenerate[] = ['key' => $p, 'file' => $p . '.html'];
        }

        $generated = [];
        foreach ($pagesToGenerate as $page) {
            $tmpl = self::$templateDir . '/' . $page['file'];
            if (!file_exists($tmpl)) continue;

            $nav  = self::buildNavLinks($pages, $page['key']);
            $html = self::interpolate(file_get_contents($tmpl), array_merge($baseVars, ['NAV_LINKS' => $nav]));
            file_put_contents($outputDir . '/' . $page['file'], $html);
            $generated[] = $page['file'];
        }

        return $generated;
    }

    private static function interpolate(string $template, array $vars): string
    {
        return preg_replace_callback('/\{\{(\w+)\}\}/', function ($m) use ($vars) {
            return $vars[$m[1]] ?? $m[0];
        }, $template);
    }

    private static function buildNavLinks(array $pages, string $currentPage): string
    {
        $active = array_merge(['index'], $pages);
        $links  = [];
        foreach (self::$allPages as $p) {
            if (!in_array($p['key'], $active, true)) continue;
            $cls = $p['key'] === $currentPage ? 'nav-link nav-active' : 'nav-link';
            $links[] = sprintf('<a href="%s" class="%s">%s</a>', $p['file'], $cls, $p['label']);
        }
        return implode("\n        ", $links);
    }
}

File: rust/src/generator.rs
use include_dir::{include_dir, Dir};
use regex::Regex;
use std::collections::HashMap;
use std::fs;
use std::path::Path;

use crate::themes::get_theme;

static TEMPLATES: Dir = include_dir!("$CARGO_MANIFEST_DIR/templates");

struct PageDef {
    key: &'static str,
    label: &'static str,
    file: &'static str,
}

const ALL_PAGES: &[PageDef] = &[
    PageDef { key: "index",     label: "🏠 Home",      file: "index.html" },
    PageDef { key: "about",     label: "👤 About Me",   file: "about.html" },
    PageDef { key: "gallery",   label: "🖼️ Gallery",    file: "gallery.html" },
    PageDef { key: "guestbook", label: "📖 Guestbook",  file: "guestbook.html" },
    PageDef { key: "links",     label: "🔗 Cool Links", file: "links.html" },
];

fn interpolate(template: &str, vars: &HashMap<&str, String>) -> String {
    let re = Regex::new(r"\{\{(\w+)\}\}").unwrap();
    re.replace_all(template, |caps: &regex::Captures| {
        vars.get(caps.get(1).unwrap().as_str())
            .cloned()
            .unwrap_or_else(|| caps.get(0).unwrap().as_str().to_string())
    })
    .into_owned()
}

fn build_nav_links(pages: &[String], current_page: &str) -> String {
    let active: Vec<&str> = std::iter::once("index")
        .chain(pages.iter().map(String::as_str))
        .collect();

    ALL_PAGES
        .iter()
        .filter(|p| active.contains(&p.key))
        .map(|p| {
            let cls = if p.key == current_page {
                "nav-link nav-active"
            } else {
                "nav-link"
            };
            format!(r#"<a href="{}" class="{}">{}</a>"#, p.file, cls, p.label)
        })
        .collect::<Vec<_>>()
        .join("\n        ")
}

pub fn generate(output_dir: &Path, answers: &crate::Answers) -> Vec<String> {
    let theme = get_theme(&answers.theme);
    let year = chrono::Local::now().format("%Y").to_string();

    fs::create_dir_all(output_dir.join("css")).unwrap();
    fs::create_dir_all(output_dir.join("js")).unwrap();

    let mut base_vars: HashMap<&str, String> = HashMap::new();
    base_vars.insert("SITE_NAME",        answers.site_name.clone());
    base_vars.insert("AUTHOR_NAME",      answers.author_name.clone());
    base_vars.insert("THEME_NAME",       theme.name.to_string());
    base_vars.insert("BG",               theme.bg.to_string());
    base_vars.insert("BG_PATTERN",       theme.bg_pattern.to_string());
    base_vars.insert("BG_SIZE",          theme.bg_size.to_string());
    base_vars.insert("TEXT_COLOR",       theme.text_color.to_string());
    base_vars.insert("HEADING_COLOR",    theme.heading_color.to_string());
    base_vars.insert("LINK_COLOR",       theme.link_color.to_string());
    base_vars.insert("LINK_HOVER",       theme.link_hover.to_string());
    base_vars.insert("ACCENT_COLOR",     theme.accent_color.to_string());
    base_vars.insert("BORDER_LIGHT",     theme.border_light.to_string());
    base_vars.insert("BORDER_DARK",      theme.border_dark.to_string());
    base_vars.insert("PANEL_BG",         theme.panel_bg.to_string());
    base_vars.insert("PANEL_BORDER",     theme.panel_border.to_string());
    base_vars.insert("COUNTER_BG",       theme.counter_bg.to_string());
    base_vars.insert("COUNTER_TEXT",     theme.counter_text.to_string());
    base_vars.insert("TABLE_BORDER",     theme.table_border.to_string());
    base_vars.insert("TABLE_HEADER_BG",  theme.table_header_bg.to_string());
    base_vars.insert("GLOW_COLOR",       theme.glow_color.to_string());
    base_vars.insert("NAV_ACTIVE_BG",    theme.nav_active_bg.to_string());
    base_vars.insert("NAV_ACTIVE_TEXT",  theme.nav_active_text.to_string());
    base_vars.insert("CURSOR_EFFECT",    answers.cursor_effect.clone());
    base_vars.insert("FALLING_EFFECT",   answers.falling_effect.clone());
    base_vars.insert("WELCOME_ALERT",    if answers.welcome_alert  { "true" } else { "false" }.to_string());
    base_vars.insert("PLAY_MUSIC",       if answers.play_music     { "true" } else { "false" }.to_string());
    base_vars.insert("FAKE_HIGH_COUNT",  if answers.fake_high_count{ "true" } else { "false" }.to_string());
    base_vars.insert("YEAR",             year);

    let css_src = TEMPLATES.get_file("css/style.css").unwrap().contents_utf8().unwrap();
    fs::write(output_dir.join("css/style.css"), interpolate(css_src, &base_vars)).unwrap();

    let js_src = TEMPLATES.get_file("js/main.js").unwrap().contents_utf8().unwrap();
    fs::write(output_dir.join("js/main.js"), interpolate(js_src, &base_vars)).unwrap();

    let pages_to_generate: Vec<(&str, &str)> = std::iter::once(("index", "index.html"))
        .chain(answers.pages.iter().map(|p| (p.as_str(), ALL_PAGES.iter().find(|x| x.key == p).map(|x| x.file).unwrap_or(""))))
        .filter(|(_, f)| !f.is_empty())
        .collect();

    let mut generated = Vec::new();
    for (key, file) in &pages_to_generate {
        if let Some(tmpl) = TEMPLATES.get_file(file) {
            let nav = build_nav_links(&answers.pages, key);
            let mut vars = base_vars.clone();
            vars.insert("NAV_LINKS", nav);
            let html = interpolate(tmpl.contents_utf8().unwrap(), &vars);
            fs::write(output_dir.join(file), html).unwrap();
            generated.push(file.to_string());
        }
    }

    generated
}
